src/languageHandler.js

```diff
@@ -64,75 +61,112 @@ export function _t(...args) {
  */
-export function _tJsx(jsxText, patterns, subs) {
-    // convert everything to arrays
-    if (patterns instanceof RegExp) {
-        patterns = [patterns];
-    }
-    if (subs instanceof Function) {
-        subs = [subs];
-    }
-    // sanity checks
-    if (subs.length !== patterns.length || subs.length < 1) {
-        throw new Error(`_tJsx: programmer error. expected number of RegExps == number of Functions: ${subs.length} != ${patterns.length}`);
-    }
-    for (let i = 0; i < subs.length; i++) {
-        if (!(patterns[i] instanceof RegExp)) {
-            throw new Error(`_tJsx: programmer error. expected RegExp for text: ${jsxText}`);
+export function substitute(text, variables, tags) {
+    const regexpMapping = {};
+
+    if(variables !== undefined) {
+        for (const variable in variables) {
+            regexpMapping[`%\\(${variable}\\)s`] = variables[variable];
         }
-        if (!(subs[i] instanceof Function)) {
-            throw new Error(`_tJsx: programmer error. expected Function for text: ${jsxText}`);
+    }
+
+    if(tags !== undefined) {
+        for (const tag in tags) {
+            regexpMapping[`(<${tag}>(.*?)<\\/${tag}>|<${tag}>|<${tag}\\s*\\/>)`] = tags[tag];
         }
     }
+    return replaceByRegexes(text, regexpMapping);
+}
 
-    // The translation returns text so there's no XSS vector here (no unsafe HTML, no code execution)
-    const tJsxText = _t(jsxText, {interpolate: false});
-    const output = [tJsxText];
+/*
+ * Replace parts of a text using regular expressions
+ * @param {string} text The text on which to perform substitutions
+ * @param {object} mapping A mapping from regular expressions in string form to replacement string or a
+ * function which will receive as the argument the capture groups defined in the regexp. E.g.
+ * { 'Hello (.?) World': (sub) => sub.toUpperCase() }
+ *
+ * @return a React <span> component if any non-strings were used in substitutions, otherwise a string
+ */
+export function replaceByRegexes(text, mapping) {
+    const output = [text];
+
+    let wrap = false; // Remember if the output needs to be wrapped later
+    for (const regexpString in mapping) {
+        const regexp = new RegExp(regexpString);
 
-    for (let i = 0; i < patterns.length; i++) {
         // convert the last element in 'output' into 3 elements (pre-text, sub function, post-text).
         // Rinse and repeat for other patterns (using post-text).
         const inputText = output.pop();
-        const match = inputText.match(patterns[i]);
-        if (!match) {
-            throw new Error(`_tJsx: translator error. expected translation to match regexp: ${patterns[i]}`);
+        const match = inputText.match(regexp);
+        if(!match) {
+            output.push(inputText); // Push back input
+            continue; // Missing matches is entirely possible, because translation might change things
```

Comment:
I thought the point was that the translations should contain the things to be replaced? Should we not still be throwing an error here?

Reply:
This is not true. You can pass in e.g. `count`, but not display it.

Reply:
If `count` is the exception, then maybe we should handle it as one and treat missing matches as errors?

Reply:
It's also possible for others, but probably less common. E.g. in the "plural branch" you might display other variables than in the "singluar branch".

Reply:
Warning would be fine, but probably want to ignore missing `count` because otherwise we log pointlessly. In doing so we might lose some warnings we do want for plural branches, as you've described, but that feels like an okay tradeoff.